Error flagging with status codes (#136)
open-telemetry/oteps: text/trace/0136-error_flagging.md

# Error Flagging with Status Codes

> This proposal adds two status codes explicitly for use as overrides by the end user, and proposes a canonical mapping of semantic conventions to status codes. This clarifies how error reporting should work in OpenTelemetry.

Reviewer: It's actually three status codes if "`UNSET` is the default status code" is included. It's also not only intended for overriding by the end user but also to be set by instrumentations.

Author: fixed

> This proposal adds two status codes explicitly for use as overrides by the end user, and proposes a canonical mapping of semantic conventions to status codes. This clarifies how error reporting should work in OpenTelemetry.

Reviewer: What is meant by the _canonical mapping of semantic conventions to status codes_?

Author: removed

## Motivation

> Error reporting is a fundamental use case for distributed tracing. While we prefer that error flagging occurs within analysis tools, and not within instrumentation plugins, a number of currently supported analysis tools and protocols rely on the existence of an explicit error flag reported from instrumentation. In OpenTelemetry, the error flag is called "status codes".

Reviewer: ```suggestion
Error reporting is a fundamental use case for distributed tracing. While we prefer that error flagging occurs within analysis tools, and not within instrumentations, a number of currently supported analysis tools and protocols rely on the existence of an explicit error flag reported from instrumentation. In OpenTelemetry, the error flag is called "status codes".
```

This would introduce a new term, usually it's simply called instrumentation in the spec.

Reviewer 2: Sorry to nitpick your nitpick, but this shouldn't be pluralized here.
```suggestion
Error reporting is a fundamental use case for distributed tracing. While we prefer that error flagging occurs within analysis tools, and not within instrumentation, a number of currently supported analysis tools and protocols rely on the existence of an explicit error flag reported from instrumentation. In OpenTelemetry, the error flag is called "status codes".
```

Author: Fixed. Officially we call them [instrumentation libraries](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/glossary.md#instrumentation-library) in the spec. I added a clarification at the beginning for the terms `end user` and `instrumentation`.

> However, there is confusion over the mapping of semantic conventions to status codes, and concern over the subjective nature of errors. Which network failures count as an error? Are 404s an error? The answer is often dependent on the situation, but without even a baseline of suggested status codes for each convention, the instrumentation author is placed under the heavy burden of making the decision. Worse, the decisions will not be in sync across different instrumentation packages.

Reviewer: ```suggestion
However, there is confusion over the mapping of semantic conventions to status codes, and concern over the subjective nature of errors. Which network failures count as an error? Are 404s an error? The answer is often dependent on the situation, but without even a baseline of suggested status codes for each convention, the instrumentation author is placed under the heavy burden of making the decision. Worse, the decisions will not be in sync across different instrumentations.
```

Yet another new term ;-)

Reviewer 2: Same
```suggestion
However, there is confusion over the mapping of semantic conventions to status codes, and concern over the subjective nature of errors. Which network failures count as an error? Are 404s an error? The answer is often dependent on the situation, but without even a baseline of suggested status codes for each convention, the instrumentation author is placed under the heavy burden of making the decision. Worse, the decisions will not be in sync across different instrumentation.
```

Reviewer 2: Though, actually, I don't like how this reads like this. Perhaps _instrumentation providers_? _Sources of instrumentation_?
Now that I'm struggling with wording here, I wonder if it's really that important.

Author: We officially call them [instrumentation libraries](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/glossary.md#instrumentation-library). I included a note at the top to clarify that this is what the term "instrumentation" refers to in this OTEP.

> There is one other missing piece, required for proper error flagging. Both application developers and operators have a deep understanding of what constitutes an error in their system. OpenTelemetry must provide a way for these users to control error flagging, and explicitly indicate that it is the end user setting the status code, and not instrumentation plugins. In these specific cases, the error flagging is known to be correct: the end user has decided the status of the span, and they do not want another interpretation.

Reviewer: ```suggestion
There is one other missing piece, required for proper error flagging. Both application developers and operators have a deep understanding of what constitutes an error in their system. OpenTelemetry must provide a way for these users to control error flagging, and explicitly indicate that it is the end user setting the status code, and not an instrumentation. In these specific cases, the error flagging is known to be correct: the end user has decided the status of the span, and they do not want another interpretation.
```

Reviewer 2: ```suggestion
There is one other missing piece, required for proper error flagging. Both application developers and operators have a deep understanding of what constitutes an error in their system. OpenTelemetry must provide a way for these users to control error flagging, and explicitly indicate that it is the end user setting the status code, and not instrumentation. In these specific cases, the error flagging is known to be correct: the end user has decided the status of the span, and they do not want another interpretation.
```

> There is one other missing piece, required for proper error flagging. Both application developers and operators have a deep understanding of what constitutes an error in their system. OpenTelemetry must provide a way for these users to control error flagging, and explicitly indicate that it is the end user setting the status code, and not instrumentation plugins. In these specific cases, the error flagging is known to be correct: the end user has decided the status of the span, and they do not want another interpretation.

Reviewer: End user would be the person _using_ the application (e.g., the person shopping on the instrumented web shop) rather than the developer instrumenting the application, right? Maybe we can call it application or instrumentation developer? I think end user only makes sense from a back end vendor's POV (like we both are) but not for a general audience.

> While generic instrumentation can only provide a generic schema, end users are capable of making subjective decisions about their systems. And, as the end user, they should get to have the final call in what constitutes an error. In order to accomplish this, there must be a way to differentiate between errors flagged by instrumentation, and errors flagged by the end user.

Reviewer: Same as above: I don't think end user is the right term to use here.

## Explanation

The following changes add several missing features required for proper error reporting, and are completely backwards compatible with OpenTelemetry today.

### Status Codes

Currently, OpenTelemetry does not have a use case for differentiating between different types of errors. However, this use case may appear in the future. For now, we would like to reduce the number of status codes, and then add them back in as the need becomes clear. We would also like to differentiate between status codes which have not been
set, and an explicit OK status set by an end user.

* `UNSET` is the default status code.

> * `ERROR` represents all error types.

Reviewer 3: Can this be named "UnknownError":
1. Consistent with status codes;
2. In the future if other errors are added this means not other more specific error was identified;

Reviewer 4: In the future if other errors are added, I would rather add another enum ErrorCategory or something like that (maybe as semantic convention). That would also be backwards compatible with code that would then check `if (status == ERROR) { assume error } else { assume no error }`.

Author: I think it should map to `UNKOWN`, but the name `UNKOWN` doesn't make sense when we only have one kind of error. When we add more codes, `ERROR` will then mean "generic error, no details" which is more or less how we treat `UNKOWN` today.

> * `OK` represents a span which has been explicitly marked as being free of errors, and should not be counted against an error budget. Note that only end users should set this status. Instead, instrumentation should leave the status as `UNSET` for normal operations.

Reviewer: ```suggestion
* `OK` represents a span which has been explicitly marked as being free of errors, and should not be counted against an error budget. Note that only end users should set this status. Instead, instrumentations should leave the status as `UNSET` for normal operations.
```

> * `OK` represents a span which has been explicitly marked as being free of errors, and should not be counted against an error budget. Note that only end users should set this status. Instead, instrumentation should leave the status as `UNSET` for normal operations.

Reviewer: What are _normal operations_?

Author: operations which do not generate an error.

Reviewer 2: ```suggestion
* `OK` represents a span which has been explicitly marked as being free of errors, and should not be counted against an error budget. Note that only end users should set this status. Instead, instrumentation should leave the status as `UNSET` for operations that do not generate an error.
```

> ### `Status Source`

Reviewer 3: Can this be added later, not sure we thought about this enough and feels like a bit rushing. Is there any downside if we do that?

Reviewer 4: I have mixed feelings about splitting the boolean from the main status code enum (see earlier versions of this PR), but I think this OTEP does not make sense without having the user-supplied-info/instrumentation-derived-heuristic split.

Author: For backends which want to ignore errors from instrumentation and instead use their own heuristic, we need a way to inform them that the status code is an override which they should respect. We could have `ERROR_INSTRUMENTATION `and `ERROR_USER`, but that would lead to a duplication of all the different error codes as we add them back in. People didn't like that part - this felt clearer. Having them all be status codes would be atomic, which is nice.

Author: (But, to be clear, I think this is a fine compromise.)

> A new Status Source field identifies the origin of the status code on the span. This is important, as statuses set by application developers and operators have been confirmed by the end user to be correct to the particular situation. Statuses set by instrumentation, on the other hand, are only following a generic schema.

Reviewer: Is the _end user_ mentioned here different from _application developers and operators_?

Author: no, it is the same.

Author: I added a the top to clarify this.

* `INSTRUMENTATION` is the default source. This is used for instrumentation contained within shared code, such as OSS libraries and frameworks. All instrumentation plugins shipped with OpenTelemetry use this status code.

> * `USER` identifies statuses set by the end user, either in application code or the collector.

Reviewer: ```suggestion
* `USER` identifies statuses set by application developers or operators, either in application code or the collector.
```

> Analysis tools MAY disregard status codes, in favor of their own approach to error analysis. However, it is strongly suggested that analysis tools SHOULD pay attention to the status codes when set by `USER`, as it is a communication from the end-user and contains valuable information.

Reviewer: ```suggestion
Analysis tools MAY disregard status codes, in favor of their own approach to error analysis. However, it is strongly suggested that analysis tools SHOULD pay attention to the status codes when set by `USER`, as it is a communication from the application developer or operator and contains valuable information.
```
The data an instrumentation provides is usually also valuable 😉

### Status Mapping Schema

> As part of the specification, OpenTelemetry provides a canonical mapping of semantic conventions to status codes. This removes any ambiguity as to what OpenTelemetry ships with out of the box.

Reviewer: I don't really understand this sentence.
Does this mean the semantic conventions will be expanded to provide a mapping of a library's/framework's return status to whether it should be `ERROR` or `OK` or something else? Can we provide a reasonable mapping anywhere other than for HTTP? Our semantic conventions target generic _concepts_ like DB calls, RPC or messaging but usually not concrete libraries (e.g. MongoDB return code X). Will we add something like that?

Author: Yes, I would like to improve these conventions over time, to ensure that instrumentation is normalized across languages and. For example, we do not want to label a MongoDB code an error in one client, but not an error in another. We should do our best to record in the spec what counts as an error in the instrumentation we provide, when those errors relate to something cross-langauge, like a database protocol. 

It's also not clear to me if "generic" concepts like DB will be sufficient in the long run. SQL, Mongo, and Redis are all very different protocols. But even if the DB conventions remains generic, we still need to describe how to map each protocol to this generic convention, to ensure the data that we emit is consistent. I would like to bring in experts from these different databases to help with this.

Reviewer 2: I think we can provide reasonable mappings for databases, though the conventions will need to include special cases (but really, have you seen the db semantic conventions? It feels like it's mostly special cases 😄). We can probably provide something pretty good for messaging RPC too.  

Also, I'd bet we can make a pretty good go at a general convention as well.  If the operation results in an exception or a panic, it's probably an error (is that controversial?).

Author: Great minds think alike :)

I added a clarification to this section.

Reviewer: I see! That sounds reasonable to me.

@tedsuo I think you forgot to push the clarification but once that's done we're good to go :shipit:

Author: @arminru derp! Pushed. :)

Obviously, all systems are different, and users will want to adjust error reporting on a case by case basis. Unwanted errors may be suppressed, and additional errors may be added. The collector will provide a processor and a configuration language to make this a straightforward process. Working from a baseline of standard errors will provide a better experience than having to define a schema from scratch.

Note that analysis tools MAY disregard Span Status, and do their own error analysis. There is no requirement that the status code is respected, even when Status Source is set. However, it is strongly suggested that analysis tools SHOULD pay attention to the status code when Status Source is set, as it represents a subjective decision made by either the operator or application developer.

## Remind me why we need status codes again?

Status codes provide a low overhead mechanism for checking if a span counts against an error budget, without having to scan every attribute and event. It is an inexpensive and low cardinality approach to track multiple types of error budgets. This reduces overhead and could be a benefit for many systems.

> However, adding in an existing set of error types without first clearly defining their use and how they might be set has caused confusion. If the status codes are not set consistently and correctly, then the resulting error budgeting will not be useful. So we are consolidating all error types into a single ERROR type, to avoid this situation. We may add more error types back in if we can agree on their use cases and a method for applying them consistently.

Reviewer 5: > We may add more **error types** back in if

(emphasis mine) be careful to refer to these as **status codes** to signal that they are (theorized to be) useful for reporting status types that are not errors, for example "canceled".

Author: Thanks, I keep catching myself doing that. :)

# Error Flagging with Status Codes

This proposal reduces the number of status codes to three, adds a new field to identify status codes set by application developers and operators, and adds a mapping of semantic conventions to status codes. This clarifies how error reporting should work in OpenTelemetry.

> Note: the term **end user** is defined as the application developers and operators of the system running opentelemetry. The term **instrumentation** refers to [instrumentation libraries](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/glossary.md#instrumentation-library) for common code shared between different systems, such as web frameworks and database clients.

Reviewer: ```suggestion
Note: The term **end user** in this document is defined as the application developers and operators of the system running OpenTelemetry. The term **instrumentation** refers to [instrumentation libraries](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/glossary.md#instrumentation-library) for common code shared between different systems, such as web frameworks and database clients.
```
